pkg/providers/version/version.go

```diff
@@ -69,39 +70,39 @@ func NewDefaultProvider(kubernetesInterface kubernetes.Interface, cache *cache.C
 }
 
 func (p *DefaultProvider) Get(ctx context.Context) (string, error) {
-	var version string
-	var versionSource string
-	if version, ok := p.cache.Get(kubernetesVersionCacheKey); ok {
-		return version.(string), nil
+	if version := *p.previousVersion.Clone(); version != "" {
+		return version, nil
 	}
-	output, err := p.eksapi.DescribeCluster(ctx, &eks.DescribeClusterInput{
-		Name: lo.ToPtr(options.FromContext(ctx).ClusterName),
-	})
-	if err != nil {
-		if !awserrors.IsAccessDenied(err) {
-			return "", err
-		}
-		output, err := p.kubernetesInterface.Discovery().ServerVersion()
-		if err != nil {
-			return "", fmt.Errorf("getting kubernetes version from the kubernetes API")
-		} else if output != nil {
-			version = fmt.Sprintf("%s.%s", output.Major, strings.TrimSuffix(output.Minor, "+"))
-			versionSource = "Kubernetes API"
-		}
-	} else if lo.FromPtr(output.Cluster.Version) != "" {
-		version = *output.Cluster.Version
+	return "", fmt.Errorf("kubernetes version not yet cached")
+}
+
+func (p *DefaultProvider) UpdateVersion(ctx context.Context) error {
+	var version, versionSource string
+	var err error
+
+	if options.FromContext(ctx).EKSControlPlane {
+		version, err = p.getEKSVersion(ctx)
 		versionSource = "EKS DescribeCluster"
 	} else {
-		return "", fmt.Errorf("unable to retrieve Kubernetes version from EKS DescribeCluster")
+		version, err = p.getK8sVersion()
+		versionSource = "Kubernetes API"
+	}
+	if version == "" && err == nil {
+		version = *p.previousVersion.Clone()
+	}
+	if err != nil {
+		log.FromContext(ctx).Error(err, "failed to get kubernetes version")
+		return err
 	}
-	p.cache.SetDefault(kubernetesVersionCacheKey, version)
+	p.previousVersion.Set(&version)
 	if p.cm.HasChanged("kubernetes-version", version) || p.cm.HasChanged("version-source", versionSource) {
 		log.FromContext(ctx).WithValues("version", version, "source", versionSource).V(1).Info("discovered kubernetes version")
 		if err := validateK8sVersion(version); err != nil {
 			log.FromContext(ctx).Error(err, "failed validating kubernetes version")
+			return err
```

Comment:
Similar to my previous comment, we should either log or return, not both since that will result in duplicate log lines (typically). If you want to add addtional context to the error, you can do so like this:
```suggestion
			return fmt.Errorf("validating kubernetes version, %w", err)
```